lib/sudoers/src/ast.rs

```diff
@@ -0,0 +1,303 @@
+use crate::basic_parser::*;
+use crate::tokens::*;
+use std::iter::Peekable;
+
+/// The Sudoers file allows negating items with the exclamation mark.
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub enum Qualified<T> {
+    Allow(T),
+    Forbid(T),
+}
+
+/// Type aliases; many items can be replaced by ALL, aliases, and negated.
+pub type Spec<T> = Qualified<Meta<T>>;
+pub type SpecList<T> = Vec<Spec<T>>;
+
+/// The RunAs specification consists of a (possibly empty) list of userspecifiers, followed by a (possibly empty) list of groups.
+#[derive(Debug, Default)]
+pub struct RunAs {
+    pub users: SpecList<UserSpecifier>,
+    pub groups: SpecList<Username>,
+}
+
+/// Commands in /etc/sudoers can have attributes attached to them, such as NOPASSWD, NOEXEC, ...
+#[allow(clippy::upper_case_acronyms)]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Tag {
+    NOPASSWD,
```

Comment:
I think having normal rust identifiers here might be better compared to the exact token as specified in the sudoers file

Reply:
Most of the allowed-to-be-dead code turned out not to be dead code anymore; I've changed the Tag::identifiers, although I think it can be good to have these objects stand out since they can be rather important and need to change the behavior of components outside of this crate. At the same time as soon as you start pattern matching on them Rustc will probably force us to deal with all possible alternatives.